python/_tskitmodule.c

```diff
@@ -9860,6 +9860,150 @@ TreeSequence_divergence_matrix(TreeSequence *self, PyObject *args, PyObject *kwd
     return ret;
 }
 
+/* TODO:
+ * in the default case (no time windows) we calculate statistics at the level of nodes.
+ * this is triggered by a zero-length time windows array, rather than None, to avoid
+ * juggling NULL pointers. Is there a better way to go about this? */
+static int
+parse_time_windows(
+    PyObject *time_windows, PyArrayObject **ret_array, tsk_size_t *ret_num_time_windows)
+{
+    int ret = -1;
+    tsk_size_t num_time_windows = 0;
+    PyArrayObject *time_windows_array = NULL;
+    npy_intp *shape;
+
+    time_windows_array = (PyArrayObject *) PyArray_FROMANY(
+        time_windows, NPY_FLOAT64, 1, 1, NPY_ARRAY_IN_ARRAY);
+    if (time_windows_array == NULL) {
+        goto out;
+    }
+    shape = PyArray_DIMS(time_windows_array);
+    if (shape[0] == 1) { /* allow zero length array */
```

Comment:
All these error cases should be provoked in ``test_lowlevel.py``. This really is critical, as it's super easy to shoot yourself in the foot in the Python C API. A small mistake handling an error condition can lead to a really obscure core dump minutes later in totally unrelated parts of the program, which can take weeks to track down. (I have a war stories on this - really!)